form-flow/src/main/kotlin/com/ritense/formflow/domain/instance/FormFlowStepInstance.kt

```diff
@@ -103,13 +104,43 @@ data class FormFlowStepInstance(
     }
 
     private fun <T> processExpressions(expressions: List<String?>): List<T?> {
-        return ExpressionProcessorFactoryHolder.getInstance().let {
+        return ExpressionProcessorFactoryHolder.getInstance().let { factory ->
             val variables = createVarMap()
-            val expressionProcessor = it.create(variables)
+            val expressionProcessor = factory.create(variables)
 
-            expressions.map { expression ->
+            val results = expressions.map { expression ->
                 expression?.let { expressionProcessor.process<T>(expression) }
             }
+            val newCompleteSubmissionData = MapperSingleton.get().valueToTree<JsonNode>(variables)
+                .at("/step/submissionData")
+            updateSubmissionData(newCompleteSubmissionData)
+            results
+        }
+    }
+
+    private fun updateSubmissionData(newSubmissionData: JsonNode) {
+        val mapper = MapperSingleton.get()
+        val oldCompleteSubmissionData = mapper.readValue<JsonNode>(instance.getSubmissionDataContext())
+        if (newSubmissionData != oldCompleteSubmissionData) {
+            keepDiff(newSubmissionData, oldCompleteSubmissionData)
+            if (this.submissionData != null) {
+                this.submissionData = newSubmissionData.toString()
+            } else {
+                this.temporarySubmissionData = newSubmissionData.toString()
+            }
+        }
+    }
+
+    private fun keepDiff(target: JsonNode, toRemove: JsonNode) {
+        if (target is ObjectNode && toRemove is ObjectNode) {
+            target.properties().toList().forEach { (key, value) ->
```

Comment:
.properties().toList() can also be fields, but more efficient: https://fasterxml.github.io/jackson-databind/javadoc/2.7/com/fasterxml/jackson/databind/JsonNode.html#fields()

Reply:
I just tried that and unfortunately I get a `ConcurrentModificationException`